File: ErmisCommon/src/github/koukobin/ermis/common/results/ChangePasswordResult.java
/* Copyright (C) 2023 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */
package github.koukobin.ermis.common.results;

import java.util.Arrays;
import java.util.HashMap;
import java.util.stream.Collectors;

import github.koukobin.ermis.common.util.EnumIntConverter;

/**
 * @author Ilias Koukovinis
 *
 */
public enum ChangePasswordResult {
    SUCCESFULLY_CHANGED_PASSWORD(0, true, "Succesfully changed password!"),
    ERROR_WHILE_CHANGING_PASSWORD(1, false, "There was an error while trying to change password!");

    private static final HashMap<Integer, ChangePasswordResult> values;

    static {
		values = new HashMap<>(
				Arrays.stream(ChangePasswordResult.values())
				.collect(Collectors.toMap(type -> type.id, type -> type))
				);
    }

    public final ResultHolder resultHolder;
    public final int id;

    ChangePasswordResult(int id, boolean isSuccesfull, String message) {
        resultHolder = new ResultHolder(isSuccesfull, message);
        this.id = id;
    }

    public static ChangePasswordResult fromId(int id) {
        return EnumIntConverter.fromId(values, id);
    }
}

File: ErmisClient/Mobile/ermis_client/lib/core/networking/common/results/change_password_result.dart
/* Copyright (C) 2024 Ilias Koukovinis <ilias.koukovinis@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 * 
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

import 'package:ermis_mobile/features/authentication/domain/entities/resultable.dart';
import 'package:ermis_mobile/generated/l10n.dart';

import '../../../exceptions/enum_not_found_exception.dart';

enum ChangePasswordResult implements Resultable {
  successfullyChangedPassword(
    id: 0,
    success: true,
  ),
  errorWhileChangingPassword(
    id: 1,
    success: false,
  );

  final int id;
  final bool success;

  const ChangePasswordResult({
    required this.id,
    required this.success,
  });

  @override
  bool get isSuccessful => success;
  
  @override
  String get message => switch(ChangePasswordResult.fromId(id)) {
    ChangePasswordResult.successfullyChangedPassword => S.current.change_password_success,
    ChangePasswordResult.errorWhileChangingPassword => S.current.change_password_error
  };

  static ChangePasswordResult fromId(int id) {
    try {
      return ChangePasswordResult.values.firstWhere((type) => type.id == id);
    } catch (e) {
      throw EnumNotFoundException('No $ChangePasswordResult found for id $id');
    }
  }
}
